### PR TITLE
CNN updates

### PR DESCRIPTION
**Pull request by LewisOs**
Hi Shane,

Here are the CNN updates. I've a strong suspicion that I'm not using the loops properly in the nested loop section, but otherwise I think the rest is alright... Hope it looks okay to you.

Cheers, Lewis.

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -79,29 +79,36 @@ def cnn_model_fn(features, labels, mode, params):
     training = mode == tf.estimator.ModeKeys.TRAIN
     # loop for adding a convolutional layer and max pooling layer pair
     for layer in params["layers"]:
+        for conv in range(layer["num_convs"]):
```
Comment (shanest) [inline models.py]: Two small comments, one very pedantic and one somewhat more interesting.

(i) Could you use single quotes / apostrophes when accessing dictionary fields (e.g. `params['layers']` and `layer['num_convs']`, and a few other places in these two files)?  It doesn't matter to a huge degree, but the consistency with how I've been doing it suits my aesthetic sensibility :-).

(ii) I would also write the nested loop as `for _ in range(layer['num_convs'])`.  Currently, you're assigning to `conv` the integer values in the `range`; but notice that you never use that variable `conv`.  Using the underscore `_` allows you to run the loop without assigning any variables that then don't get used.

File: models.py [hunk 2]
```diff
@@ -79,29 +79,36 @@ def cnn_model_fn(features, labels, mode, params):
...
             pool_size=layer['pool_size'],
             strides=layer['strides'])
 
-    # flattening the output of last max pooling layer into a batch of vectors
+    # flattening the output of last max pooling layer into a batch of vectors    
     net = tf.reshape(net, [batch_size, np.prod(net.shape[1:])])
 
     # Dense Layer
-    net = tf.layers.dense(inputs=net, units=1024, activation=tf.nn.relu)
+    for layer in params["dense"]:
```
Comment (shanest) [inline models.py]: See earlier comment on single quotes :-)

**Review by shanest — CHANGES_REQUESTED** [review]
Hi Lewis,

The code looks great!  I'm requesting a few changes before merging the pull request, largely just to exhibit the mechanics of a code review :-).  You should see some comments on the files in this pull request.

There are two other changes that I'd request:

(1) You should add "and Lewis O'Sullivan" to the copyright statement at the top of the files you've worked on :-).

(2) Can you delete the non-Python files from the repository? (e.g. .DS_Store, __pycache__, images/.DS_Store). You can delete them from the web interface for GitHub; then, in the future when using the Desktop app, you can just make sure that they are unchecked when you make a commit.

Cheers!

**Comment by LewisOs** [comment]
Hi Shane,

I've done the edits you requested. Hope it all looks okay.

Ta, Lewis.

File: models.py [hunk 3]
```diff
@@ -171,6 +178,24 @@ def glimpse_network(images, locs, scope='glimpse_network'):
...
             # params['g_size'] and params['l_size']
+            hidden_sensor_layer = tf.layers.dense(
+                    # not sure this is the correct input
+                    inputs = patches,
+                    units = params['g_size'],
+                    activation = tf.nn.relu)
+            
+            hidden_location_layer = tf.layers.dense(
+                    # unsure of input here
+                    inputs = ,
```
Comment (shanest) [inline models.py:189]: `inputs` should be `locs`, which is of shape `[batch_size, 2]`

**Comment by shanest** [comment]
Hi Lewis,

Code looks great!  And well done getting on the glimpse_network too!

Two very, very minor things before I finally pull :-):

* inputs should be `locs` to that layer of the glimpse network you were wondering about

* Somehow this update added some Spyder files and a directory "NNs for most" to the repository.  Can you delete those?  (Remember to always check which files are checked as being committed in the Desktop app.)

Once these little things are done, I'll pull it in to the main repo!

**Comment by LewisOs** [comment]
Hi again -  I think all the extra files are gone now. I'll send another pull request when I'm done working on the glimpse network.

Cheers, Lewis.

**Review by shanest — APPROVED** [review]

**Comment by shanest** [comment]
Awesome, I just merged into the main repo.  Congrats on your first successful pull request!!

**Pull request merged**